WalletWasabi/Wallets/FilterProcessor/BlockDownloadService.cs

```diff
@@ -0,0 +1,276 @@
+{
+	/// <summary>Maximum number of parallel block-downloading tasks.</summary>
+	private const int MaxParallelTasks = 5;
+
+	/// <summary>Maximum number of attempts to download a block. If it fails, we drop the block download request altogether.</summary>
+	public const int MaxFailedAttempts = 3;
+
+	public BlockDownloadService(IFileSystemBlockRepository fileSystemBlockRepository, IBlockProvider blockProvider, int maximumParallelTasks = MaxParallelTasks)
+	{
+		FileSystemBlockRepository = fileSystemBlockRepository;
+		BlockProvider = blockProvider;
+		MaximumParallelTasks = maximumParallelTasks;
+	}
+
+	/// <summary>Source that provided a block or blocks.</summary>
+	public enum Source
+	{
+		FileSystemCache,
+		BlockProvider
+	}
+
+	/// <summary>Result object describing if/how object was downloaded using the block downloading service.</summary>
+	public interface IResult { }
+
+	/// <summary>Denotes a failure of getting a block.</summary>
+	public interface IFailureResult : IResult { }
+
+	private IFileSystemBlockRepository FileSystemBlockRepository { get; }
+
+	/// <remarks>Implementation must provide caching functionality - i.e. once a block is downloaded, it must be readily available next time it is requested.</remarks>
+	private IBlockProvider BlockProvider { get; }
+	private int MaximumParallelTasks { get; }
+	private SemaphoreSlim SynchronizationRequestsSemaphore { get; } = new(initialCount: 0, maxCount: 1);
+
+	/// <summary>Block hashes that are to be downloaded. Block height represents priority of the priority queue.</summary>
+	/// <remarks>
+	/// Guarded by <see cref="Lock"/>.
+	/// <para>Internal for testing purposes.</para>
+	/// </remarks>
+	internal PriorityQueue<Request, Priority> BlocksToDownload { get; } = new(Priority.Comparer);
+
+	/// <remarks>Guards <see cref="BlocksToDownload"/>.</remarks>
+	private object Lock { get; } = new();
+
+	/// <summary>
+	/// Add a block hash to the queue to be downloaded.
+	/// </summary>
+	public TaskCompletionSource<IResult> Enqueue(uint256 blockHash, Priority priority) =>
+		Enqueue(new Request(blockHash, priority, 1, new TaskCompletionSource<IResult>()));
+
+	private TaskCompletionSource<IResult> Enqueue(Request request)
+	{
+		lock (Lock)
+		{
+			int count = BlocksToDownload.Count;
+			BlocksToDownload.Enqueue(request, request.Priority);
+
+			if (count == 0 && SynchronizationRequestsSemaphore.CurrentCount == 0)
+			{
+				SynchronizationRequestsSemaphore.Release();
+			}
+		}
+
+		return request.Tcs;
+	}
+
+	/// <summary>
+	/// Remove all blocks whose height is larger or equal to <paramref name="maxBlockHeight"/>.
+	/// </summary>
+	/// <remarks>
+	/// The method is not efficient but it does not matter too much because the operation is not supposed to be called often.
+	/// The main use case is to deal with a blockchain reorg.
+	/// </remarks>
+	public void RemoveBlocks(uint maxBlockHeight)
+	{
+		PriorityQueue<Request, Priority> tempQueue = new(Priority.Comparer);
```

Comment:
You might be able to solve this as well without locks. You just pass a Reorg "command" into the channel and done it in the loop.